cleverhans/attacks.py

```diff
@@ -1,4 +1,313 @@
+from abc import ABCMeta, abstractmethod
 import numpy as np
+import warnings
+
+from .utils import random_targets
+
+
+class Attack:
+    """
+    Abstract base class for all attack classes. All attacks must override the
+    `craft` method that returns adversarial examples corresponding to the input
+    data. However, the `generate_symbolic` method is optional.
+    """
+    __metaclass__ = ABCMeta
+
+    def __init__(self, x, pred, back='tf', sess=None, clip_min=None,
+                 clip_max=None, params={}):
+        """
+        :param x: The model's symbolic inputs.
+        :param pred: The model's symbolic output.
+        :param back: The backend to use. Either 'tf' (default) or 'th'.
+        :param clip_min: Minimum float value for adversarial example components
+        :param clip_max: Maximum float value for adversarial example components
+        :param params: Parameter dictionary used by child classes.
+        """
+        if not back == 'tf' or back == 'th':
+            raise Exception("Backend argument must either be 'tf' or 'th'.")
+        if back == 'tf' and sess is None:
+            raise Exception("A tf session was not provided in sess argument.")
+        if back == 'th' and sess is not None:
+            raise Exception("A session should not be provided when using th.")
+        self.x = x
+        self.pred = pred
+        self.back = back
+        self.sess = sess
+        self.clip_min = clip_min
+        self.clip_max = clip_max
+
+    def generate_symbolic(self):
```

Comment:
This would make sense if we had a symbolic implementation of every attack. But saliency map attack doesn't (so we can't just do sess.run after computing a symbolic version).